apps/frontend/src/layout/PublicRoute.tsx

```diff
@@ -0,0 +1,68 @@
+import { GlobalSpinner } from "@/components";
+import { User, useViewerLazyQuery } from "@/generated/graphql";
+import { useAuthToken } from "@/modules/auth/hooks/useAuthToken";
+import useRedirect from "@/modules/auth/hooks/useRedirect";
+import useSession from "@/modules/users/hooks/useSession";
+import { Session } from "@/modules/users/states/session";
+import { omitTypename } from "@/utils";
+import { logDebug } from "@/utils/log";
+import React, { useEffect, useState } from "react";
+
+export default function PublicRoute({
+    children,
+}: {
+    children: React.ReactNode;
+}) {
+    const [ready, setReady] = useState(false);
+    const { token } = useAuthToken();
+    const { session, isValidSession, user, createSession, updateUser } =
+        useSession();
+    const [fetchUser] = useViewerLazyQuery();
+    const redirect = useRedirect();
+
+    useEffect(() => {
+        if (!token) {
+            setReady(true);
+            return;
+        }
+
+        if (!user) {
+            logDebug("No user, fetching user...");
+            fetchUser({
+                onCompleted: ({ viewer }) => {
+                    if (viewer) {
+                        logDebug("Gotten user, updating session...");
+                        const newUser = omitTypename(viewer as User);
+                        updateUser(newUser);
+                        createSession({
+                            organization: newUser?.organization,
+                            project: newUser?.project,
+                        } as Session);
+                    }
+                },
+            });
+        }
+
+        if (!session && user) {
```

Comment:
@AfeezGL I think with the new `useSession`, we can get rid of most of the logic here? Is my assumption correct?

Reply:
Yes.